### DISCUSSION TITLE
useDebounceFetcher and returned action data

### DISCUSSION THREAD

**Discussion by caprica** [sergiodxa/remix-utils#306] [Q&A] [answered]
What's the idiomatic way to get returned data from useDebounceFetcher?

My use-case is to display error messages that may be returned from the action.

**Comment by sergiodxa** [comment, marked as answer]
The response will have the data in `fetcher.data` like a `useFetcher`.

**Comment by caprica** [comment]
Sorry, my question was a bit a vague, I think it is more related to the typing, which is maybe where my confusion comes from.

If I have an action like this:

```typescript
export const action = () => {
   // ... whatever ...
   return json({ firstName, lastName})
}
```

Isn't this normal to type the fetcher?

```typescript
const fetcher = useFetcher<typeof action>();
const data = fetcher.data
```

`data` is of this type:

```typescript
const data: JsonifyObject<{
    firstName: string;
    lastName: string;
}> | undefined
```

With debounceFetcher:

```typescript
const fetcher = useDebounceFetcher<typeof action>();
const data = fetcher.data
```

This time `data` looks like this:

```typescript
const data: (({ request }: ActionFunctionArgs) => Promise<{
    firstName: string;
    lastName: string;
}>) | undefined
```

Which is the type of the action function itself, not what it returns.

So to use debounceFetcher with Typescript, I have to declare explicitly a new type rather than using typeof<action> I suppose.

This would work I think:

```typescript
const fetcher = useDebounceFetcher<Awaited<ReturnType<typeof action>>>()
```

**Reply by luisrudge** [reply]
oh, I thought I was going crazy 😅 I'm having the same issue here. @sergiodxa any ideas on how to address that? thanks!

**Reply by luisrudge** [reply]
`const fetcher = useDebounceFetcher<Awaited<ReturnType<typeof action>>>()`
this doesn't work because if you use `json()` in your action, the return type is actually `TypedResponse<T>`
<img width="772" alt="image" src="https://github.com/user-attachments/assets/02108613-dd41-4a11-93d6-d73c04d97382">

**Reply by luisrudge** [reply]
```
export type UpdateSubmissionActionResult = Awaited<ReturnType<Awaited<ReturnType<typeof action>>['json']>>;
```

this works, but it's pretty ugly 😅

**Reply by luisrudge** [reply]
if someone is having the same issue, you can easily fix this when `singleFetch` is enabled

```
export type UpdateSubmissionActionResult = SerializeFrom<typeof action>;
```

maybe it works without singleFetch too

**Reply by zlach** [reply]
I also came across this issue; just going to type it manually

`const fetcher = useDebounceFetcher<{myThing: string}>()`
